Concat required libs

samvimes01 opened the pull request:
- add `mode` field to `build.json` with `app`, `iife` and default `lib` modes
- add `libDir` field to `build.json` to treat cases like `metacom/dist`
- skip `import` lines alongside with `require`
- refactor - split code to separate concerns, improved naming, update readme
- add config validation with metaschema
- add various tests
- changed filename comment to `#region` comment to allow collapse of code region in vscode-like editors

https://github.com/metarhia/metautil/issues/311

<!--
Thank you for your pull request.
Check following steps to help us land your changes:
Change [ ] to [x] for completed items.
-->

- [x] tests and linter show no problems (`npm t`)
- [x] tests are added/updated for bug fixes and new features
- [x] code is properly formatted (`npm run fix`)
- [ ] description of changes is added in CHANGELOG.md
- [ ] update .d.ts typings

tshemsedinov commented on metarhia-build.js:
```diff
@@ -54,13 +54,22 @@ const build = () => {
     `// Version ${packageJson.version} ${packageName} ${licenseName}\n\n`;
+  const requiredLibs = (buildConfig.require || [])
+    .map((lib) => {
+      const fileName = `./node_modules/${lib}/${lib}.mjs`;
```
Do not build dependencies from node_modules

samvimes01 replied:
I've added iife build mode (separate commit - easy to revert).
This will include deps from node_modules to prepare lib for worker. We will need it for metacom in serviceworker

samvimes01 replied:
Applied changes - as we discussed in telegram @tshemsedinov 

After Firefox fixed bug with service worker imports https://bugzilla.mozilla.org/show_bug.cgi?id=1360870

Changed the builder
- lib mode builds `lib.mjs` file and adds metarhia deps imports like `import metautil from './metautil.js'` - `.js`
- then in app mode symlinks `.mjs` as `.js` so end user app will fnd correct esm files served from static folder

Rebuild corresponding metacom and matautil PRs
metacom - (to master branch) https://github.com/metarhia/metacom/pull/528
metautil - (to created by you `build` branch) https://github.com/metarhia/metautil/pull/315/changes
also there is a metachema - (to master) https://github.com/metarhia/metaschema/pull/480

tshemsedinov reviewed (commented):
Build steps in short
1. npm library build
  1.1. Remove require/import of local sub-modules
  1.2. Convert `module.exports` to export
  1.3. Concatenate library files but not dependencies
2. application build
  2.1. Add links to `node_modules` files from server `/static/` folder
  2.2. Alternatively application can use any bundler to build ours npm files in traditional way

samvimes01 commented:
I've changed how deps are processed. No need for `require` field in build json.
If it's `node:` deps - lib build will skip this line and logs warning.
For `node_modules` deps it will find all such deps and put `import from` only once at the top of the bundle.

I've added `mode` field to `build.json`
1. New `app` mode.
When builder is used in the app with app mode it will link libs listed in the `order` field to` application/static` folder

2. New `iife` mode.
When lib is required to run inside worker context where ES modules unavailable.
This wraps lib in iife and concat it with `node_modules` deps instead of importing such deps. (workers throws errors when spot `import` or `export` keyword)
